crowbar_framework/lib/crowbar/upgrade_status.rb

```diff
@@ -75,13 +75,19 @@ def current_step_state
       progress[:steps][current_step] || {}
     end
 
-    def start_step
+    # 'step' is name of the step user wants to start.
+    # If `out_of_order` is true, the step could be executed out of order
+    def start_step(step_name, out_of_order = false)
       ::Crowbar::Lock::LocalBlocking.with_lock(shared: false, logger: @logger, path: lock_path) do
-        if running?
+        if running? step_name
```

Comment:
This is mostly to prevent one step for being started twice in parallel. But you are right it does not exactly prevent us to having 2 steps running at the same time (if one of them is `out_of_order`). We might allow that (as @skazi0 suggests), or we might extend the `running?` method to check if *any* step is currently running.